agent/pkg/agent/docker_provider.go

```diff
@@ -0,0 +1,446 @@
+	entries, err := os.ReadDir(containerProcPath)
+	if err != nil {
+		return 0, err
+	}
+
+	inodeStr := fmt.Sprintf("socket:[%d]", inode)
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		pid, err := strconv.Atoi(entry.Name())
+		if err != nil {
+			continue
+		}
+
+		fdDir := fmt.Sprintf("%s/%d/fd", containerProcPath, pid)
+		fdEntries, err := os.ReadDir(fdDir)
+		if err != nil {
+			continue
+		}
+
+		for _, fdEntry := range fdEntries {
+			linkPath := filepath.Join(fdDir, fdEntry.Name())
+			target, err := os.Readlink(linkPath)
+			if err != nil {
+				continue
+			}
+			if target == inodeStr {
+				return pid, nil
+			}
+		}
+	}
+
+	return 0, fmt.Errorf("no process found for inode %d", inode)
+}
+
+// readFileInContainer reads a file from container's /proc (deprecated, use readFileFromProc)
+func readFileInContainer(path string) string {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return "unknown"
+	}
+	return strings.TrimSpace(string(data))
+}
+
+// readFileFromProc reads a file from /proc filesystem
+func readFileFromProc(path string) string {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return "unknown"
+	}
+	return strings.TrimSpace(string(data))
+}
+
+// findProcessesInNamespace finds all processes in the given network namespace
+func findProcessesInNamespace(targetNs string) ([]int, error) {
+	entries, err := os.ReadDir("/proc")
+	if err != nil {
+		return nil, fmt.Errorf("failed to read /proc: %w", err)
+	}
+
+	var pids []int
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		pid, err := strconv.Atoi(entry.Name())
+		if err != nil {
+			continue // Not a PID directory
+		}
+
+		// Check if this process is in the target namespace
+		nsPath := fmt.Sprintf("/proc/%d/ns/net", pid)
+		ns, err := os.Readlink(nsPath)
+		if err != nil {
+			continue // Process may have exited or no permission
+		}
+
+		if ns == targetNs {
+			pids = append(pids, pid)
+		}
+	}
+
+	if len(pids) == 0 {
+		return nil, fmt.Errorf("no processes found in namespace %s", targetNs)
+	}
+
+	return pids, nil
+}
+
+// findProcessByInodeInNamespace finds a process by inode within a set of PIDs
+func findProcessByInodeInNamespace(pids []int, inode uint64) (int, error) {
+	inodeStr := fmt.Sprintf("socket:[%d]", inode)
+
+	for _, pid := range pids {
+		fdDir := fmt.Sprintf("/proc/%d/fd", pid)
+		fdEntries, err := os.ReadDir(fdDir)
+		if err != nil {
+			continue // Process may have exited or no permission
+		}
+
+		for _, fdEntry := range fdEntries {
+			linkPath := filepath.Join(fdDir, fdEntry.Name())
+			target, err := os.Readlink(linkPath)
+			if err != nil {
+				continue
+			}
+			if target == inodeStr {
+				return pid, nil
+			}
+		}
+	}
+
+	return 0, fmt.Errorf("no process found for inode %d in namespace", inode)
+}
+
+// isDockerIP checks if an IP belongs to common Docker network ranges
+func isDockerIP(ipStr string) bool {
```

Comment:
I guess it doesn't work for non-default network ranges? And so the worst case is that we report `docker` as the process that was blocked, and not a process name within docker?

Reply:
Fixed in https://github.com/bullfrogsec/bullfrog/pull/216/commits/d1d875500a16f2bf37e907b617887d03fde497e6